packages/data-broker/src/services/dhis/DhisInputSchemeResolvingApi.js

```diff
@@ -0,0 +1,306 @@
+  // SS162: { dhisId: 'pDhCn9lbi5v' },
+  // SS128: { dhisId: 'K3wJsRJHMCa' },
+  // SS182: { dhisId: 'g2KqfrYXLdM' },
+  // SS190A: { dhisId: 'iYzYjNzDnAY' },
+  // SS192: { dhisId: 'RngB7fLGI3i' },
+  // SS219: { dhisId: 'bcg0ApKZTav' },
+  // DP9: { dhisId: 'OzDb71SXJgP' },
+  HP02: { dhisId: 'OmssK27WEJB' },
+};
+
+export class DhisInputSchemeResolvingApi {
+  constructor(models, api) {
+    this.models = models;
+    this.api = api;
+  }
+
+  async getAnalytics(query) {
+    let modifiedQuery = { ...query };
+
+    // console.log('DhisInputSchemeResolvingApi', 'getAnalytics', 'query', query);
+
+    if (
+      (await this.anyDataElementsHaveDhisId(query)) ||
+      (await this.anyOrgUnitsHaveDhisId(query))
+    ) {
+      // the endpoint used /api/analytics/rawData.json only allows a single "inputIdScheme", which means
+      // both the dataElements and orgUnits need to be ids
+
+      modifiedQuery = await this.replaceDataElementCodesWithIds(modifiedQuery);
+      modifiedQuery = await this.replaceOrgUnitCodesWithIds(modifiedQuery);
+
+      modifiedQuery.inputIdScheme = 'uid';
+    }
+
+    // console.log('DhisInputSchemeResolvingApi', 'getAnalytics', 'modifiedQuery', modifiedQuery);
+    return this.api.getAnalytics(modifiedQuery);
+  }
+
+  async getEventAnalytics(query) {
+    // console.log('DhisInputSchemeResolvingApi', 'getEventAnalytics', 'query', query);
+    let modifiedQuery = { ...query };
+
+    const originalDataElementIdScheme = query.dataElementIdScheme;
+
+    const areDataElementsUsingDhisId = await this.anyDataElementsHaveDhisId(query);
+
+    if (areDataElementsUsingDhisId) {
+      modifiedQuery = await this.replaceDataElementCodesWithIds(modifiedQuery);
+      modifiedQuery.dataElementIdScheme = 'id';
+    }
+
+    if (await this.anyProgramsHaveDhisId(query)) {
+      modifiedQuery = await this.replaceProgramCodesWithIds(modifiedQuery);
+    }
+
+    if (await this.anyOrgUnitsHaveDhisId(query)) {
+      modifiedQuery = await this.replaceOrgUnitCodesWithIds(modifiedQuery);
+    }
+
+    // console.log('DhisInputSchemeResolvingApi', 'getEventAnalytics', 'modifiedQuery', modifiedQuery);
+
+    const response = await this.api.getEventAnalytics(modifiedQuery);
+
+    if (areDataElementsUsingDhisId && originalDataElementIdScheme === 'code') {
+      // The api response will contain data elements with ids, and DhisApi will not be able to translate these
+      // back into codes (because the codes are not set in dhis).
+      // So, we have to do it ourselves using the internal mapping
+
+      const dataElementIdToCode = {};
+
+      // const dataElements = await this.models.dataSource.find({ id: query.dataElementCodes });
+      //
+      // for (const dataElement of dataElements) {
+      //   if (dataElement.config.dhisId) {
+      //     dataElementIdToCode[dataElement.dhisId] = dataElement.code;
+      //   }
+      // }
+
+      for (const dataElementCode of query.dataElementCodes) {
```

Comment:
This conversion back allows data builders to work with codes instead of ids, for known mappings,